### DISCUSSION TITLE
Custom link formatter

### DISCUSSION THREAD

**Discussion by aeharding** [remarkjs/remark#1179] [Q&A] [answered]
Hi there!

I have a string like this: `!hello@example.com` (a Lemmy community link).

I need to parse this, and transform into a link.

However, whenever I run my plugin alongside remark-gfm, it runs before it and parses as `mailto:` before I can do anything.

Ordering of plugins doesn't matter (I use `ReactMarkdown`).

Any help would be GREATLY appreciated! I just want my plugin to run before remark gfm parses stuff.

```html
<ReactMarkdown remarkPlugins={[myCustomPlugin, remarkGfm]}>check out !community@lemmy.world</ReactMarkdown>
```

**Comment by ChristianMurphy** [comment, marked as answer]
Welcome @aeharding!

For context see the `unified` architecture. (https://github.com/unifiedjs/unified#description)
`remark-gfm` is a parser, it will always run before a transformer plugin, which is what I assume `myCustomPlugin` is.

You could create your own parser plugin, but it is advisable not to for user/author experience purposes, see https://github.com/micromark/micromark#extending-markdown
As an alternative, consider using directives, which are semi-standard and pre-built https://github.com/remarkjs/remark-directive

**Reply by aeharding** [reply]
Hi @ChristianMurphy, thanks for the response!

I looked at directives, but I believe they need to be a certain format? I am parsing a link format that was already decided on, so I have no option there.

I ended up forking [mdast-util-gfm-autolink-literal](https://github.com/aeharding/mdast-util-gfm-autolink-literal/commit/39953aed8d2461c3f7d1ee57dfaf823dab518af5) so it doesn't parse emails if preceded by a `!`. It works... just a bit ugly 😅

You can see my super hacky custom plugin here: https://github.com/aeharding/wefwef/pull/28/files#diff-a2f4cdf9b98b1aed4fc2ddcdffe6c8a397690a356fa6bccb478d74bf9648e7b3

Thanks again for your help. And any more tips on a better approach appreciated!

**Reply by ChristianMurphy** [reply]
> I looked at directives, but I believe they need to be a certain format?

Yes, directives have a specific format.

> I am parsing a link format that was already decided on, so I have no option there.

Looking at Lemmy's docs, they don't appear to support GFM https://join-lemmy.org/docs/users/02-media.html
They are mostly commonmark based with a few extensions like sub/super (https://github.com/Symbitic/remark-plugins/blob/master/packages/remark-supersub/README.md) and spoilers.
So you may not want or need to add the GFM plugin in the first place?

**Reply by aeharding** [reply]
Hi @ChristianMurphy! Thanks, I've considered it! However, I need table support and support for inline links (e.g. https://example.com without markdown formatting) as well.

Do you know if there exists a remarkjs plugin for autolinking text links that runs as a transformer? Just something simple, I don't need GFM compliance as you pointed out.

**Reply by ChristianMurphy** [reply]
> Do you know if there exists a remarkjs plugin for autolinking text links that runs as a transformer?

I'm not aware of one, but checking the full list of community plugins may turn up something up: https://github.com/topics/remark-plugin

**Reply by wooorm** [reply]
you can wrap https://github.com/syntax-tree/hast-util-find-and-replace and pass it a regex
